cgi-bin/DW/External/ProfileServices.pm

```diff
@@ -0,0 +1,175 @@
+    my $uid = $u->userid;
+
+    # load accounts from memcache
+    my $memkey   = [ $uid, "profile_accts:$uid" ];
+    my $accounts = LJ::MemCache::get($memkey);
+    return $accounts if $accounts;
+
+    $accounts = {};
+
+    # load accounts from database and add to memcache (no expiration)
+    my $dbcr = LJ::get_cluster_reader($u) or die;
+    my $data = $dbcr->selectall_arrayref(
+        "SELECT name, value FROM user_profile_accts WHERE userid=? ORDER BY name, value",
+        { Slice => {} }, $uid );
+    die $dbcr->errstr if $dbcr->err;
+
+    foreach my $acct (@$data) {
+        my $name = $acct->{name};
+        $accounts->{$name} //= [];
+        push @{ $accounts->{$name} }, $acct->{value};
+    }
+
+    LJ::MemCache::set( $memkey, $accounts );
+
+    return $accounts;
+}
+*LJ::User::load_profile_accts = \&load_profile_accts;
+*DW::User::load_profile_accts = \&load_profile_accts;
+
+sub save_profile_accts {
+    my ( $u, $new_accts, %opts ) = @_;
+    $u = LJ::want_user($u) or confess 'invalid user object';
+    my $old_accts = $u->load_profile_accts;
+
+    # expire memcache after updating db
+    my $uid    = $u->userid;
+    my $memkey = [ $uid, "profile_accts:$uid" ];
+
+    return unless $u->writer;
+
+    my %sites = map { $_->{name} => 1 } @{ DW::External::ProfileServices->list };
+
+    # if %$old_accts is empty, we need to clear out the user's legacy userprops
+    # to avoid an edge case where if a user clears out all of their accounts
+    # later, the old userprop values will suddenly reappear on their profile
+    unless (%$old_accts) {
+        my $userprops = DW::External::ProfileServices->userprops;
+        my %prop      = map { $_ => '' } @$userprops;
+        $u->set_prop( \%prop, undef, { skip_db => 1 } );
+    }
+
+    # convert listref vals to sanitized hashref vals
+    # note: we now lowercase to avoid value duplication
+    # when allowing multiple values per account type
+    my $remap = sub {
+        my $href = shift;
+        my %ret;
+
+        foreach my $name ( keys %$href ) {
+            next unless $sites{$name};
+            foreach my $val ( @{ $href->{$name} } ) {
+                $val = LJ::text_trim( lc( $val // '' ) );
+                next unless $val;
+                $ret{$name} //= {};
+                $ret{$name}->{$val} = 1;
+            }
+        }
+        return %ret;
+    };
+
+    my %old = $remap->($old_accts);
+    my %new = $remap->($new_accts);
+
+    my $process = sub {
+        my ( $hr1, $hr2 ) = @_;
+        my %ret;
+
+        foreach my $name ( keys %$hr1 ) {
+            foreach my $val ( keys %{ $hr1->{$name} } ) {
+                unless ( $hr2->{$name} && $hr2->{$name}->{$val} ) {
+                    $ret{$name} //= [];
+                    push @{ $ret{$name} }, $val;
+                }
+            }
+        }
+        return %ret;
+    };
+
+    my %del = $process->( \%old, \%new );
+    my %add = $process->( \%new, \%old );
```

Comment:
this should probably enforce some kind of limit to how many external services you can link?

Reply:
Right now, the limit is enforced by having a static number of form fields on /manage/profile as shown in the screenshot - I think it's 26. @momijizukamori mentioned wanting to make that section of the form dynamic in the future, so we can come back and decide what the future limit should be at that point.